### ISSUE TITLE
Embedding function with tabular potentials

### ISSUE THREAD

**Issue by ymishin-gmu** [potfit/potfit#108]
The common practice with EAM and ADP potentials is to enforce the embedding function F(rho) to be exactly zero at rho=0. Even if a rho=0 configuration is not in the training DFT dataset, it may occur during the subsequent MD simulations (e.g., deposition from vapor). 

For analytical potentials, the F(rho=0)=0 condition is enforced by properly choosing the functions. However, I couldn't figure out how to enforce this condition for a tabulated potential. When I set this condition in the initial potential, the optimization (using evo) changes F(rho) to nonzero.

Maybe I am doing something wrong. Please help me figure this out.

Thank you,
Yuri Mishin.

**Comment by dschopf** [comment]
Hi @ymishin-gmu, sorry for the delayed response. As far as I know there is no such thing as fixing individual data points for tabulated potentials in potfit at the moment. Even though I have to admit that your use case seems quite reasonable. 

From a technical point of view this would be quite some effort to implement in the current code as it is not prepared for such a feature. As a workaround maybe you could simply reset the value after each change back to 0, that should be possible in a few lines of code.

**Comment by ymishin-gmu** [comment]
Thank you for getting back to me, Daniel. Resetting to zero would be a reasonable temporary solution. I would rather not tamper with the code, but if the developers or users implement this option at some point in the future, that will be great, and I will use it. Thank you, -- Yuri.

**Comment by p-bro** [comment]
Just to add to this: I think one could get away with limited changes around https://github.com/potfit/potfit/blob/8e2479f75597a4cb1eb3dc31d24e85c0344593dd/src/potential_input_f3.c#L210C7-L210C7 (for type 3 files with equidistant sampling). This if statement includes the potential value in the list of parameters to optimise if yes, and skips it if false. If one would add a condition such as `!(pt->xcoord[l]==0. && pt->table[l]==0.)` (made floating-point safe, so comparing if abs smaller some delta) to that line, it would skip entries where rho=0 and F=0 in the optimisation, keeping them zero.

**Comment by dschopf** [comment]
I have implemented the suggestion from Peter because my suggestion is most likely not feasible. It requires lots of changes in several places because of the architecture of potfit.

Please check this branch https://github.com/potfit/potfit/tree/eamFZeroRhoZero and use the option `--enable-fzerorhozero`for enabling this feature. This is highly experimental and I did only verify that this code compiles. If you have some time please test and provide feedback if this is what you are looking for.

**Comment by ghost** [comment]
I tested the code for tabulated potential in EAM/ADP, compiling it with the --enable-fzerorhozero option. A zero initial value for embedding function F(rho) at rho = 0 produces zero. However, when a non-zero value is used, it does not produce zero for rho = 0. I would appreciate any suggestions you may have. 

Thank you,
Indiras Khatri

**Comment by dschopf** [comment]
@Kindiras yes that is how it works. If the initial value is zero then it will be excluded from the optimization, otherwise it will be optimized as before. Is this not what you expected or need?

**Comment by ymishin-gmu** [comment]
Yes Daniel, this solves the problem for us. Many thanks for addressing our request.

From: Daniel Schopf ***@***.***>
Date: Tuesday, January 30, 2024 at 2:27 PM
To: potfit/potfit ***@***.***>
Cc: Yuri Mishin ***@***.***>, Mention ***@***.***>
Subject: Re: [potfit/potfit] Embedding function with tabular potentials (Issue #108)

@Kindiras<https://github.com/Kindiras> yes that is how it works. If the initial value is zero then it will be excluded from the optimization, otherwise it will be optimized as before. Is this not what you expected or need?
